File: sdks/typescript/src/evaluators/inputs.ts
import { InputValidationError } from '../errors.js';

/**
 * Validating a caller's inputs against the schema the evaluator declares.
 *
 * Each evaluator's `input_schema.json` is the only description of what it accepts, so
 * the bounds, the accepted grades and the set of field names all come from there. A
 * global default applied to every evaluator would silently ignore a contract asking for
 * something narrower, which §4.1 forbids.
 */

/** The shape of an `input_schema.json`, as much of it as validation reads. */
export interface DeclaredInputSchema {
  properties: Record<
    string,
    { type?: string; minLength?: number; maxLength?: number; enum?: string[] }
  >;
  required?: string[];
}

/**
 * The inputs an evaluator accepts, derived from its declared schema.
 *
 * Keys come from the schema, so adding an input to a contract is a compile error in
 * every call site that does not pass it. Values are `string` because TypeScript widens
 * imported JSON string literals — the declared `enum` and bounds are enforced at
 * runtime by {@link validateInputs}, which is the authoritative check either way.
 */
export type InputsOf<S extends { properties: object }> = Record<keyof S['properties'], string>;

/**
 * Check `inputs` against `schema`, in the order §4.1 fixes.
 *
 * Fields are visited in declared order — `required` first, then any remaining
 * properties — so a caller passing two bad inputs always gets the same message, in
 * this SDK and in any other reading the same schema.
 *
 * @throws {InputValidationError} On an unknown key, a missing field, a whitespace-only
 * or out-of-bounds string, or a value outside a declared `enum`.
 */
export function validateInputs(
  inputs: Record<string, unknown>,
  schema: DeclaredInputSchema,
): void {
  // A JS caller, or an `any`, can hand over something that is not an object at all.
  // Reaching Object.keys with it would raise a TypeError, which is neither diagnosable
  // nor part of the taxonomy.
  if (inputs === null || typeof inputs !== 'object' || Array.isArray(inputs)) {
    throw new InputValidationError(
      `Expected an object of inputs, received ${inputs === null ? 'null' : typeof inputs}.`,
    );
  }

  const declared = Object.keys(schema.properties);

  // Contracts declare `additionalProperties: false`, so an unexpected key is a caller
  // mistake worth naming rather than something to quietly drop.
  for (const key of Object.keys(inputs)) {
    if (!declared.includes(key)) {
      throw new InputValidationError(
        `Unknown input "${key}". This evaluator accepts: ${declared.join(', ')}.`,
      );
    }
  }

  const required = schema.required ?? [];
  const order = [...required, ...declared.filter((f) => !required.includes(f))];

  for (const field of order) {
    const spec = schema.properties[field];
    const value = inputs[field];

    if (value === undefined || value === null) {
      if (required.includes(field)) {
        throw new InputValidationError(`${field} is required.`);
      }
      continue;
    }

    if (spec.type === 'string' && typeof value !== 'string') {
      throw new InputValidationError(`${field} must be a string.`);
    }

    if (typeof value === 'string') {
      validateStringField(field, value, spec);
    }
  }
}

function validateStringField(
  field: string,
  value: string,
  spec: { minLength?: number; maxLength?: number; enum?: string[] },
): void {
  if (spec.enum) {
    if (!spec.enum.includes(value)) {
      throw new InputValidationError(
        `Invalid ${field} "${value}". Accepted values: ${spec.enum.join(', ')}.`,
      );
    }
    return;
  }

  // Trimming decides whether the value is blank and nothing more: the bounds below
  // measure the string as the caller sent it, which is also what reaches the model.
  if (!value.trim()) {
    throw new InputValidationError(`${field} cannot be empty or contain only whitespace`);
  }

  if (spec.minLength !== undefined && value.length < spec.minLength) {
    throw new InputValidationError(
      `${field} is too short. Minimum length is ${spec.minLength} characters.`,
    );
  }

  if (spec.maxLength !== undefined && value.length > spec.maxLength) {
    throw new InputValidationError(
      `${field} is too long. Maximum length is ${spec.maxLength} characters.`,
    );
  }
}

/**
 * The field a report or telemetry treats as the primary text.
 *
 * The wire still carries one text length, so an evaluator with two texts has to pick
 * one: the first declared string input that is not an enum. Interim — Q-10 replaces the
 * single length with a per-input map.
 */
export function primaryTextField(schema: DeclaredInputSchema): string | undefined {
  return Object.keys(schema.properties).find((field) => {
    const spec = schema.properties[field];
    return spec.type === 'string' && !spec.enum;
  });
}

File: sdks/python/src/learning_commons_evaluators/evaluators/inputs.py
"""Validating a caller's inputs against the schema the evaluator declares (SDK spec §4).

Each evaluator's ``input_schema.json`` is the only description of what it accepts, so the
bounds, the accepted grades and the set of field names all come from there. The SDK
applies no defaults of its own.
"""

from __future__ import annotations

from collections.abc import Mapping
from typing import Any

from learning_commons_evaluators.errors import InputValidationError


def validate_inputs(inputs: Any, schema: Mapping[str, Any]) -> dict[str, str]:
    """Check ``inputs`` against ``schema`` in the order §4.1 fixes, returning canonical values.

    Fields are visited in declared order — ``required`` first, then any remaining
    properties — so a caller passing two bad inputs always gets the same message, in this
    SDK and in any other reading the same schema.

    Values come back as the strings the prompt binds: an ``int`` passed for an enumerated
    string input such as ``grade_level`` is the idiomatic Python convenience (§2.3) and is
    rendered as its token before the enum check.

    :raises InputValidationError: On a non-mapping, an unknown key, a missing field, a
        wrongly typed value, a whitespace-only or out-of-bounds string, or a value outside
        a declared ``enum``.
    """
    if not isinstance(inputs, Mapping):
        raise InputValidationError(
            f"Expected a mapping of inputs, received {type(inputs).__name__}."
        )

    properties: Mapping[str, Mapping[str, Any]] = schema.get("properties", {})
    declared = list(properties)

    # Contracts declare ``additionalProperties: false``, so an unexpected key is a caller
    # mistake worth naming rather than something to quietly drop.
    for key in inputs:
        if key not in declared:
            raise InputValidationError(
                f'Unknown input "{key}". This evaluator accepts: {", ".join(declared)}.'
            )

    required = list(schema.get("required", []))
    order = [*required, *(f for f in declared if f not in required)]

    values: dict[str, str] = {}
    for field in order:
        spec = properties[field]
        value = inputs.get(field)
        if value is None:
            if field in required:
                raise InputValidationError(f"{field} is required.")
            continue
        values[field] = _validate_field(field, value, spec)
    return values


def _validate_field(field: str, value: Any, spec: Mapping[str, Any]) -> str:
    kind = spec.get("type")
    if kind == "string":
        if isinstance(value, bool) or not isinstance(value, (str, int)):
            raise InputValidationError(f"{field} must be a string.")
        if isinstance(value, int):
            if "enum" not in spec:
                raise InputValidationError(f"{field} must be a string.")
            value = str(value)
        _validate_string(field, value, spec)
        return value
    if kind == "integer":
        if isinstance(value, bool) or not isinstance(value, int):
            raise InputValidationError(f"{field} must be an integer.")
        minimum, maximum = spec.get("minimum"), spec.get("maximum")
        if minimum is not None and value < minimum:
            raise InputValidationError(f"{field} must be at least {minimum}.")
        if maximum is not None and value > maximum:
            raise InputValidationError(f"{field} must be at most {maximum}.")
        return str(value)
    return str(value)


def _validate_string(field: str, value: str, spec: Mapping[str, Any]) -> None:
    accepted = spec.get("enum")
    if accepted is not None:
        if value not in accepted:
            raise InputValidationError(
                f'Invalid {field} "{value}". Accepted values: {", ".join(map(str, accepted))}.'
            )
        return

    # Trimming decides whether the value is blank and nothing more: the bounds below
    # measure the string as the caller sent it, which is also what reaches the model.
    if not value.strip():
        raise InputValidationError(f"{field} cannot be empty or contain only whitespace")

    minimum, maximum = spec.get("minLength"), spec.get("maxLength")
    if minimum is not None and len(value) < minimum:
        raise InputValidationError(f"{field} is too short. Minimum length is {minimum} characters.")
    if maximum is not None and len(value) > maximum:
        raise InputValidationError(f"{field} is too long. Maximum length is {maximum} characters.")


def primary_text_field(schema: Mapping[str, Any]) -> str | None:
    """The field a log line or telemetry treats as the primary text.

    The wire still carries one text length, so an evaluator with two texts has to pick
    one: the first declared string input that is not an enum.
    """
    for name, spec in schema.get("properties", {}).items():
        if spec.get("type") == "string" and "enum" not in spec:
            return name
    return None


__all__ = ["primary_text_field", "validate_inputs"]
